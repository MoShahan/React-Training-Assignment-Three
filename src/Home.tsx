import { Link } from "react-router-dom";
import "./styles/Home.css"

const Home = () => {
  return (
    <>
      <h1 className="homeTitle">Assignment 3</h1>
      <div className="homeButtonContainer">
        <Link to="/v1">
          <button className="toV1">Version 1</button>
        </Link>

        <Link to="/v2">
          <button className="toV2">Version 2</button>
        </Link>
      </div>
    </>
  );
};

export default Home;
