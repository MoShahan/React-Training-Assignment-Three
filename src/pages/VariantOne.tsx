import axios from "axios";
import { createContext, useCallback, useEffect, useState } from "react";
import MyPagination from "../components/MyPagination";
import Navbar from "../components/Navbar";
import TableList from "../components/TableList";
import useDebouncedValue from "../hooks/useDebouncedValue";
import useLocalSearchText from "../hooks/useLocalSearchText";
import { StoryType } from "../types";
import { CircularProgress } from "@mui/material";
import { MAIN_API } from "../variables";
import "../styles/VariantOne.css";

export const DeleteContext1 = createContext<any>(null);

const VariantOne = () => {

  const [pageNumber, setPageNumber] = useState(1);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(false);
  const [stories, setStories] = useState<Array<StoryType>>([]);

  const [searchText, setSearchText]: any = useLocalSearchText();

  const debouncedUrl = useDebouncedValue(
    MAIN_API + "query=" + searchText + "&page=" + (pageNumber - 1)
  );

  const handleSearchTextChange = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    setSearchText(event.target.value);
  };

  useEffect(() => {
    setStories([]);
  }, [debouncedUrl]);

  useEffect(() => {
    setLoading(true);
    setError(false);
    axios({
      method: "GET",
      url: debouncedUrl,
    })
      .then((res) => {
        setStories([...new Set<any>(res.data.hits)]);
        setLoading(false);
      })
      .catch((e: any) => {
        setError(true);
        setLoading(false);
      });
  }, [debouncedUrl]);

  const handleDeleteClick = useCallback(
    (objectId: number) => {
      const filteredStories = stories.filter((story: any) => {
        return story.objectID !== objectId;
      });
      setStories(filteredStories);
    },
    [stories]
  );

  return (
    <div className="variantOne">
      <Navbar searchText={searchText} onChange={handleSearchTextChange} variant={1}/>
      {loading ? (
        <CircularProgress
          size={100}
          sx={{
            padding: "200px",
            margin: "auto",
            color:"white"
          }}
        />
      ) : error ? (
        <h1>ERROR...</h1>
      ) : (
        <div className="variantOneData">
          <DeleteContext1.Provider value={{ onClickDelete: handleDeleteClick }}>
            <TableList stories={stories} />
          </DeleteContext1.Provider>
          <MyPagination pageNumber={pageNumber} setPageNumber={setPageNumber} />
        </div>
      )}
    </div>
  );
};

export default VariantOne;
