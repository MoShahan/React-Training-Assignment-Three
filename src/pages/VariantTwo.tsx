import { createContext, useCallback, useEffect, useRef, useState } from "react";
import Navbar from "../components/Navbar";
import useDebouncedValue from "../hooks/useDebouncedValue";
import useLocalSearchText from "../hooks/useLocalSearchText";
import { MAIN_API } from "../variables";
import axios from "axios";
import { StoryType } from "../types";
import { CircularProgress } from "@mui/material";
import TableList from "../components/TableList";
import "../styles/VariantTwo.css";

export const DeleteContext2 = createContext<any>(null);
export const LastStoryContext = createContext<any>(null);

const VariantTwo = () => {
  const [pageNumber, setPageNumber] = useState(1);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(false);
  const [stories, setStories] = useState<Array<StoryType>>([]);

  const [searchText, setSearchText]: any = useLocalSearchText();

  const debouncedUrl = useDebouncedValue(
    MAIN_API + "query=" + searchText + "&page=" + pageNumber
  );

  const handleSearchTextChange = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    setSearchText(event.target.value);
  };

  useEffect(() => {
    setStories([]);
  }, [searchText]);

  useEffect(() => {
    setLoading(true);
    setError(false);
    axios({
      method: "GET",
      url: debouncedUrl,
    })
      .then((res) => {
        setStories((prevStories: any) => {
          return [...new Set<any>([...prevStories, ...res.data.hits])];
        });
        setLoading(false);
      })
      .catch((e: any) => {
        setError(true);
      });
  }, [debouncedUrl]);

  const observer = useRef<IntersectionObserver | null>();
  const lastStory = useCallback(
    (node: any) => {
      if (loading) return;
      if (observer.current) observer.current.disconnect();
      observer.current = new IntersectionObserver(
        (entries) => {
          if (entries[0].isIntersecting && pageNumber < 50) {
            setPageNumber((prev) => prev + 1);
          }
        },
        { rootMargin: "-200px" }
      );
      if (node) observer.current?.observe(node);
    },
    [loading]
  );

  const handleDeleteClick = useCallback(
    (objectId: number) => {
      const filteredStories = stories.filter((story: any) => {
        return story.objectID !== objectId;
      });
      setStories(filteredStories);
    },
    [stories]
  );

  return (
    <div className="variantTwo">
      <Navbar
        searchText={searchText}
        onChange={handleSearchTextChange}
        variant={2}
      />
      {error ? (
        <h1>ERROR...</h1>
      ) : (
        <div className="variantTwoData">
          <DeleteContext2.Provider value={{ onClickDelete: handleDeleteClick }}>
            <LastStoryContext.Provider value={lastStory}>
              <TableList stories={stories} />
            </LastStoryContext.Provider>
          </DeleteContext2.Provider>
        </div>
      )}
      {loading ? (
        <CircularProgress
          size={100}
          sx={{
            padding: "100px",
            margin: "auto",
            color: "white",
          }}
        />
      ) : (
        ""
      )}
    </div>
  );
};

export default VariantTwo;
