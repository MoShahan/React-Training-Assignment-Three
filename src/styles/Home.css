.homeButtonContainer {
    display: flex;
    justify-content: center;
    align-items: center;
    background-color: blueviolet;
    width: 500px;
    margin: auto;
    height: 80vh;
}

.homeButtonContainer button {
    padding: 20px;
    border-radius: 15px;
    transition: all 0.5s;
    background-color: white;
    color: black;
    cursor: pointer;
    margin: 50px;
}

.homeButtonContainer button:hover {
    transform: scale(1.5);
    background-color: black;
    color: white;
}

.homeTitle {
    text-align: center;
    padding: 10px;
    margin: 10px;
}