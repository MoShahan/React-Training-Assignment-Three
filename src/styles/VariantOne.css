.variantOne {
    background-color: aquamarine;
    /* display: flex;
    flex-direction: column;
    height: max-content; */
    justify-content: center;
    align-items: center;
    text-align: center;
    min-height: 100vh;
}

.variantOneData {
    display: flex;
    justify-content: center;
    align-items: center;
    flex-direction: column;
    gap: 2rem;
    padding-bottom: 50px;
}