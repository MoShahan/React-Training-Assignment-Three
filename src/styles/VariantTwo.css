.variantTwo {
    background-color: rgba(255, 255, 255, 0.102);
    justify-content: center;
    align-items: center;
    text-align: center;
    min-height: 100vh;
    padding-bottom: 300px
}

.variantTwoData {
    display: flex;
    justify-content: center;
    align-items: center;
    flex-direction: column;
    gap: 2rem;
    padding-bottom: 50px;
}