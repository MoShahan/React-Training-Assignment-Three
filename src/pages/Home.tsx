import { Link } from "react-router-dom";
import "../styles/Home.css";

const Home = () => {
  return (
    <>
      <h1 className="homeTitle">Assignment 3</h1>
      <div className="homeButtonContainer">
        <Link to="/v1" className="linktags">
          <button className="toV1">Variant 1</button>
        </Link>

        <Link to="/v2" className="linktags">
          <button className="toV2">Variant 2</button>
        </Link>
      </div>
    </>
  );
};

export default Home;
