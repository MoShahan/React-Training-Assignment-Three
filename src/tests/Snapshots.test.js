import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import Home from "../Home";
import VariantOne from "../pages/VariantOne";
import VariantTwo from "../pages/VariantTwo";

describe("SNAPSHOT testing", () => {
  test("", () => {
    const { HomeComponent } = render(
      <BrowserRouter>
        <Home />
      </BrowserRouter>
    );
    expect(HomeComponent).toMatchSnapshot();
  });
  test("Variant 1...", () => {
    const { VariantOneComp } = render(<VariantOne />);
    expect(VariantOneComp).toMatchSnapshot();
  });

  test("Variant 2...", () => {
    const { VariantTwoComp } = render(
      <BrowserRouter>
        <VariantTwo />
      </BrowserRouter>
    );
    expect(VariantTwoComp).toMatchSnapshot();
  });
});
