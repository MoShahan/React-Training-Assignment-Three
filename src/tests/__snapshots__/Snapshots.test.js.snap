// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SNAPSHOT testing Home... 1`] = `undefined`;

exports[`SNAPSHOT testing Story... 1`] = `undefined`;

exports[`SNAPSHOT testing Variant 1... 1`] = `undefined`;

exports[`SNAPSHOT testing Variant 2... 1`] = `undefined`;
